include/hipSYCL/sycl/libkernel/detail/vec.hpp

```diff
@@ -78,20 +78,82 @@ template<> struct logical_vector_op_result<detail::dp_float>
 
 template <int...> struct vector_index_sequence {};
 
+namespace emulated_types {
+
+using uchar = unsigned char;
+using ushort = unsigned short;
+using uint = unsigned int;
+using longlong = long long;
+using ulong = unsigned long;
+using ulonglong = unsigned long long;
+
+template <class T, int Dim> struct required_alignment {
+#if defined(HIPSYCL_PLATFORM_CUDA)
+  constexpr static size_t value = alignof(T) * Dim;
+#else
+  // Use default alignment otherwise
+  constexpr static size_t value = 0;
+#endif
+};
+
+// TODO: This should probably be moved to a backend-specific location.
+#if defined(HIPSYCL_PLATFORM_CUDA)
+template <class T> struct required_alignment<T, 3>
+{ constexpr static size_t value = alignof(T); };
+template <> struct required_alignment<long, 4>
+{ constexpr static size_t value = 16; };
+template <> struct required_alignment<ulong, 4>
+{ constexpr static size_t value = 16; };
+template <> struct required_alignment<longlong, 4>
+{ constexpr static size_t value = 16; };
+template <> struct required_alignment<ulonglong, 4>
+{ constexpr static size_t value = 16; };
+template <> struct required_alignment<double, 4>
+{ constexpr static size_t value = 16; };
+#endif
+
+template <class T, int Dim>
+constexpr auto required_alignment_v = required_alignment<T, Dim>::value;
+
 template <class T, int Dim> struct emulated_vector {};
 
 template <class T>
-struct emulated_vector<T,1> { T x; };
+struct alignas(required_alignment_v<T, 1>) emulated_vector<T,1> { T x; };
 
 template <class T>
-struct emulated_vector<T,2> { T x; T y; };
+struct alignas(required_alignment_v<T, 2>) emulated_vector<T,2> { T x; T y; };
 
 template <class T>
-struct emulated_vector<T,3> { T x; T y; T z; };
+struct alignas(required_alignment_v<T, 3>) emulated_vector<T,3> { T x; T y; T z; };
 
 template <class T>
-struct emulated_vector<T,4> { T x; T y; T z; T w; };
+struct alignas(required_alignment_v<T, 4>) emulated_vector<T,4> { T x; T y; T z; T w; };
+
+// This is a hack for the host device where float1,2,3,4 etc don't exist...
+// TODO: Reimplement vec<> properly
+#define HIPSYCL_DEFINE_EMULATE_VECTOR_TYPES(T)                                 \
```

Comment:
Looks like a missing d, e.g. `EMULATE*D*_VECTOR_TYPES`? Typo?

EDIT: Oopsie, I see it's already like this in the original code.